src/goto-programs/goto_cfg.cpp

```diff
@@ -165,3 +194,646 @@ void goto_cfg::dump_graph() const
+                                   const CFGNode &bb) {
+    // Are we dealing with a global?
+    if (!symbol_map.count(var.as_string()))
+      return;
+    ;
+    symbol_information &sym = symbol_map[var.as_string()];
+
+    sym.def_blocks.insert(bb);
+    sym.def_instructions.insert(
+      std::make_shared<goto_programt::instructiont>(*I));
+  };
+
+  const auto insert_use = [&symbol_map, &unpromotable_symbols](
+                            const expr2tc &e,
+                            goto_programt::instructionst::iterator &I,
+                            const CFGNode &bb) {
+    std::unordered_set<expr2tc, irep2_hash> symbols;
+    get_addr_symbols(e, symbols);
+    for (const expr2tc &sym_expr : symbols)
+    {
+      assert(is_symbol2t(sym_expr));
+      const symbol2t &symbol = to_symbol2t(sym_expr);
+      unpromotable_symbols.insert(symbol.thename.as_string());
+    }
+
+    symbols.clear();
+    get_symbols(e, symbols);
+    for (const expr2tc &sym_expr : symbols)
+    {
+      assert(is_symbol2t(sym_expr));
+      const symbol2t &symbol = to_symbol2t(sym_expr);
+      assert(symbol_map.count(symbol.thename.as_string()));
+      symbol_information &sym = symbol_map[symbol.thename.as_string()];
+
+      // TODO: When using live analysis, only consider a use-block if the value is used before redefined
+      sym.use_blocks.insert(bb);
+      sym.use_instructions.insert(
+        std::make_shared<goto_programt::instructiont>(*I));
+    }
+  };
+
+  goto_cfg::foreach_bb(
+    start,
+    [this, &insert_definition, &insert_use, &symbol_map](const CFGNode &bb) {
+      for (goto_programt::instructionst::iterator it = bb->begin; it != bb->end;
+           it++)
+      {
+        if (it->is_decl())
+        {
+          symbol_information info;
+          const code_decl2t &decl = to_code_decl2t(it->code);
+          info.type = decl.type;
+          symbol_map.insert({decl.value.as_string(), info});
+          const symbolt *const symbol = context.find_symbol(decl.value);
+          assert(symbol != nullptr);
+          info.mode = symbol->mode.as_string();
+
+          // TODO: I am assuming that all declarations are followed by
+          //       an init assignment. So decls definitions are useless
+        }
+
+        else if (it->is_assign())
+        {
+          const expr2tc &target = to_code_assign2t(it->code).target;
+          if (is_symbol2t(target))
+            insert_definition(to_symbol2t(target).thename, it, bb);
+
+          insert_use(to_code_assign2t(it->code).source, it, bb);
+        }
+        else if (it->is_function_call())
+          for (const auto &op : to_code_function_call2t(it->code).operands)
+            insert_use(op, it, bb);
+        else
+        {
+          insert_use(it->code, it, bb);
+          insert_use(it->guard, it, bb);
+        }
+      }
+    });
+
+  for (auto symbol : unpromotable_symbols)
+    symbol_map.erase(symbol);
+
+  // TODO: Technically, not all address uses needs to be avoided
```

Comment:
The existing TODO comments are useful, but they could be made more specific.

Reply:
Not sure how much more specific I can be.